File: Stack/Problem3 - 394.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        String s = "3[a2[c]]";
        String result = decodeString(s);
        System.out.println(result);
    }

    public static String decodeString(String s) {
        int n = s.length();

        Stack<Character> st = new Stack<>();
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) != ']') {
                st.push(s.charAt(i));
            } else {
                StringBuilder currStr = new StringBuilder();
                while (!st.isEmpty() && st.peek() != '[') {
                    currStr.insert(0, st.pop());
                }
                st.pop(); // pop out open bracket 

                StringBuilder num = new StringBuilder();
                while (!st.isEmpty() && Character.isDigit(st.peek())) {
                    num.insert(0, st.pop());
                }

                int noOfTime = Integer.parseInt(num.toString());
                StringBuilder decodedStr = new StringBuilder();
                for (int j = 0; j < noOfTime; j++) {
                    decodedStr.append(currStr);
                }

                for (char ch : decodedStr.reverse().toString().toCharArray()) {
                    st.push(ch);
                }
            }
        }

        StringBuilder ans = new StringBuilder();
        while (!st.isEmpty()) {
            ans.insert(0, st.pop());
        }
        return ans.toString();
    }
}

File: Stack/Problem3 - 394.cpp
#include <bits/stdc++.h>
using namespace std;

/* optimal appraoach - using stack */
string decodeString(string s) {
    int n = s.size();

    stack<char> st;
    for (int i = 0; i < n; i++) {
        if (s[i] != ']') st.push(s[i]);
        else {
            string curr_str = "";
            while (!st.empty() && st.top() != '[') {
                curr_str = st.top() + curr_str;
                st.pop();
            }
            st.pop(); // pop out open bracket 

            string num = "";
            while (!st.empty() && (st.top() >= '0' && st.top() <= '9')) {
                num = st.top() + num;
                st.pop();
            }

            int noOfTime = stoi(num);
            while (noOfTime--) {
                for (int j = 0; j < curr_str.size(); j++) {
                    st.push(curr_str[j]);
                }
            }
        }
    }

    string ans = "";
    while (!st.empty()) {
        ans = st.top() + ans;
        st.pop();
    }
    return ans;
}